src/tools/perf/perftest_daemon.c

```diff
@@ -96,6 +107,288 @@ ucp_perf_daemon_server_conn_handle_cb(ucp_conn_request_h conn_request,
+    ucs_status_ptr_t sptr;
+
+    param.op_attr_mask = UCP_OP_ATTR_FIELD_MEMH      |
+                         UCP_OP_ATTR_FIELD_CALLBACK  |
+                         UCP_OP_ATTR_FIELD_USER_DATA |
+                         UCP_OP_ATTR_FIELD_FLAGS     |
+                         UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+    param.memh         = ctx->send_memh;
+    param.cb.send      = ucp_perf_daemon_send_cb;
+    param.user_data    = ctx;
+    param.flags        = UCP_AM_SEND_FLAG_RNDV;
+
+    sptr = ucp_am_send_nbx(ctx->peer_ep, UCP_PERF_DAEMON_AM_ID_PEER_OP, NULL, 0ul,
+                           (void*)req->addr, (size_t)req->length, &param);
+
+    return UCS_PTR_STATUS(sptr);
+}
+
+static void ucp_perf_daemon_recv_cb(void *request, ucs_status_t status,
+                                    size_t length, void *user_data)
+{
+    ucp_perf_daemon_context_t *ctx = user_data;
+
+    ucp_perf_daemon_send_am_eager_reply(ctx->host_ep,
+                                        UCP_PERF_DAEMON_AM_ID_HOST_RECV_ACK);
+    ucp_request_free(request);
+}
+
+static void ucp_perf_daemon_create_peer_ep(ucp_perf_daemon_context_t *ctx,
+                                           const void *address,
+                                           uint16_t address_length)
+{
+    ucp_ep_params_t ep_params;
+    ucs_status_t status;
+
+    memcpy(&ctx->peer_address, address, address_length);
+
+    ep_params.field_mask       = UCP_EP_PARAM_FIELD_FLAGS       |
+                                 UCP_EP_PARAM_FIELD_SOCK_ADDR   |
+                                 UCP_EP_PARAM_FIELD_ERR_HANDLER |
+                                 UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+    ep_params.err_mode         = UCP_ERR_HANDLING_MODE_PEER;
+    ep_params.err_handler.cb   = ucp_perf_daemon_err_cb;
+    ep_params.err_handler.arg  = ctx;
+    ep_params.flags            = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER;
+    ep_params.sockaddr.addr    = (struct sockaddr*)&ctx->peer_address;
+    ep_params.sockaddr.addrlen = sizeof(ctx->peer_address);
+
+    status = ucp_ep_create(ctx->worker, &ep_params, &ctx->peer_ep);
+    if (status != UCS_OK) {
+        ucs_fatal("daemon failed to create an endpoint: %s",
+                  ucs_status_string(status));
+    }
+
+    ucp_perf_daemon_send_am_eager_reply(ctx->peer_ep,
+                                        UCP_PERF_DAEMON_AM_ID_PEER_INIT);
+}
+
+static void ucp_perf_daemon_check_am_msg(const ucp_am_recv_param_t *param,
+                                         size_t header_length,
+                                         int check_reply_ep,
+                                         int check_rndv,
+                                         ucp_perf_daemon_am_id_t am_id)
+{
+    ucs_trace_data("message received: %s", ucp_perf_daemon_am_id_name(am_id));
+
+    if (check_rndv) {
+        if (!(param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV)) {
+            ucs_fatal("am message received with unsupported eager protocol");
+        }
```

Comment:
```suggestion
        ucs_assertv(param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV,
                    "am message received with unsupported eager protocol");
```

Reply:
Found better option to handle both cases:
```
    ucs_assertv(check_rndv == !!(param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV),
                "am message received with unsupported %s protocol",
                (check_rndv ? "eager" : "rndv"));
```